# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/thebes/gfxUserFontSet.h
# Commit: 43764851d067
# Full Hash: 43764851d067746870a65001786502244d7cf896
# Author: Jonathan Kew <jkew@mozilla.com>
# Date: 2021-04-07 03:19:44
# Regressor Bug: 1694123
# File Overlap Count: 1
# Description:
#   Bug 1694123 - Pass userfont source index to font loader and decoder tasks, to avoid potential race with the field in the gfxUserFontEntry being reset on the main thread. r=layout-reviewers,emilio
#   
#   This is a preliminary step that is needed in order for the following patch to be safe. (It's
#   mostly just plumbing, although the patch looks big because we need to pass the index through
#   so many functions.)
# ==============================================================================

diff -r 9b111c26978a -r 43764851d067 gfx/thebes/gfxUserFontSet.h
--- a/gfx/thebes/gfxUserFontSet.h	Tue Apr 06 16:20:47 2021 +0000
+++ b/gfx/thebes/gfxUserFontSet.h	Tue Apr 06 16:23:17 2021 +0000
@@ -266,8 +266,8 @@
       const nsTArray<mozilla::gfx::FontVariation>& aVariationSettings,
       uint32_t aLanguageOverride, gfxCharacterMap* aUnicodeRanges,
       mozilla::StyleFontDisplay aFontDisplay, RangeFlags aRangeFlags,
-      float aAscentOverride, float aDescentOverride,
-      float aLineGapOverride, float aSizeAdjust) = 0;
+      float aAscentOverride, float aDescentOverride, float aLineGapOverride,
+      float aSizeAdjust) = 0;
 
   // creates a font face for the specified family, or returns an existing
   // matching entry on the family if there is one
@@ -308,7 +308,7 @@
   // initialize the process that loads external font data, which upon
   // completion will call FontDataDownloadComplete method
   virtual nsresult StartLoad(gfxUserFontEntry* aUserFontEntry,
-                             const gfxFontFaceSrc* aFontFaceSrc) = 0;
+                             uint32_t aSrcIndex) = 0;
 
   // generation - each time a face is loaded, generation is
   // incremented so that the change can be recognized
@@ -514,7 +514,7 @@
 
   // report a problem of some kind (implemented in nsUserFontSet)
   virtual nsresult LogMessage(gfxUserFontEntry* aUserFontEntry,
-                              const char* aMessage,
+                              uint32_t aSrcIndex, const char* aMessage,
                               uint32_t aFlags = nsIScriptError::errorFlag,
                               nsresult aStatus = NS_OK) = 0;
 
@@ -649,8 +649,8 @@
   void SetLoader(nsFontFaceLoader* aLoader) { mLoader = aLoader; }
   nsFontFaceLoader* GetLoader() const { return mLoader; }
   gfxFontSrcPrincipal* GetPrincipal() const { return mPrincipal; }
-  uint32_t GetSrcIndex() const { return mSrcIndex; }
-  void GetFamilyNameAndURIForLogging(nsACString& aFamilyName, nsACString& aURI);
+  void GetFamilyNameAndURIForLogging(uint32_t aSrcIndex,
+                                     nsACString& aFamilyName, nsACString& aURI);
 
   gfxFontEntry* Clone() const override {
     MOZ_ASSERT_UNREACHABLE("cannot Clone user fonts");
@@ -663,6 +663,12 @@
 
   const nsTArray<gfxFontFaceSrc>& SourceList() const { return mSrcList; }
 
+  // Returns a weak reference to the requested source record, which is owned
+  // by the gfxUserFontEntry.
+  const gfxFontFaceSrc& SourceAt(uint32_t aSrcIndex) const {
+    return mSrcList[aSrcIndex];
+  }
+
   // The variation-query APIs should not be called on placeholders.
   bool HasVariations() override {
     MOZ_ASSERT_UNREACHABLE("not meaningful for a userfont placeholder");
@@ -698,34 +704,37 @@
   // aDownloadStatus == NS_OK ==> download succeeded, error otherwise
   // Ownership of aFontData is passed in here; the font set must
   // ensure that it is eventually deleted with free().
-  void FontDataDownloadComplete(const uint8_t* aFontData, uint32_t aLength,
-                                nsresult aDownloadStatus,
+  void FontDataDownloadComplete(uint32_t aSrcIndex, const uint8_t* aFontData,
+                                uint32_t aLength, nsresult aDownloadStatus,
                                 nsIFontLoadCompleteCallback* aCallback);
 
   // helper method for creating a platform font
   // returns true if platform font creation successful
   // Ownership of aFontData is passed in here; the font must
   // ensure that it is eventually deleted with free().
-  bool LoadPlatformFontSync(const uint8_t* aFontData, uint32_t aLength);
+  bool LoadPlatformFontSync(uint32_t aSrcIndex, const uint8_t* aFontData,
+                            uint32_t aLength);
 
-  void LoadPlatformFontAsync(const uint8_t* aFontData, uint32_t aLength,
+  void LoadPlatformFontAsync(uint32_t aSrcIndex, const uint8_t* aFontData,
+                             uint32_t aLength,
                              nsIFontLoadCompleteCallback* aCallback);
 
   // helper method for LoadPlatformFontAsync; runs on a background thread
   void StartPlatformFontLoadOnBackgroundThread(
-      const uint8_t* aFontData, uint32_t aLength,
+      uint32_t aSrcIndex, const uint8_t* aFontData, uint32_t aLength,
       nsMainThreadPtrHandle<nsIFontLoadCompleteCallback> aCallback);
 
   // helper method for LoadPlatformFontAsync; runs on the main thread
   void ContinuePlatformFontLoadOnMainThread(
-      const uint8_t* aOriginalFontData, uint32_t aOriginalLength,
-      gfxUserFontType aFontType, const uint8_t* aSanitizedFontData,
-      uint32_t aSanitizedLength, nsTArray<OTSMessage>&& aMessages,
+      uint32_t aSrcIndex, const uint8_t* aOriginalFontData,
+      uint32_t aOriginalLength, gfxUserFontType aFontType,
+      const uint8_t* aSanitizedFontData, uint32_t aSanitizedLength,
+      nsTArray<OTSMessage>&& aMessages,
       nsMainThreadPtrHandle<nsIFontLoadCompleteCallback> aCallback);
 
   // helper method for LoadPlatformFontSync and
   // ContinuePlatformFontLoadOnMainThread; runs on the main thread
-  bool LoadPlatformFont(const uint8_t* aOriginalFontData,
+  bool LoadPlatformFont(uint32_t aSrcIndex, const uint8_t* aOriginalFontData,
                         uint32_t aOriginalLength, gfxUserFontType aFontType,
                         const uint8_t* aSanitizedFontData,
                         uint32_t aSanitizedLength,
@@ -736,8 +745,8 @@
   void FontLoadFailed(nsIFontLoadCompleteCallback* aCallback);
 
   // store metadata and src details for current src into aFontEntry
-  void StoreUserFontData(gfxFontEntry* aFontEntry, bool aPrivate,
-                         const nsACString& aOriginalName,
+  void StoreUserFontData(gfxFontEntry* aFontEntry, uint32_t aSrcIndex,
+                         bool aPrivate, const nsACString& aOriginalName,
                          FallibleTArray<uint8_t>* aMetadata,
                          uint32_t aMetaOrigLen, uint8_t aCompression);
 
@@ -773,7 +782,7 @@
 
   RefPtr<gfxFontEntry> mPlatformFontEntry;
   nsTArray<gfxFontFaceSrc> mSrcList;
-  uint32_t mSrcIndex;  // index of loading src item
+  uint32_t mCurrentSrcIndex;  // index of src item to be loaded next
   // This field is managed by the nsFontFaceLoader. In the destructor and
   // Cancel() methods of nsFontFaceLoader this reference is nulled out.
   nsFontFaceLoader* MOZ_NON_OWNING_REF